dbt/include/global_project/macros/materializations/table.sql

```diff
@@ -1,36 +1,63 @@
-
 {% macro dbt__simple_create_table(schema, identifier, dist, sort, sql) -%}
     create table "{{ schema }}"."{{ identifier }}"
       {{ dist }} {{ sort }} as (
         {{ sql }}
     );
 {%- endmacro %}
 
-{% materialization table %}
+{% materialization table, adapter='base' %}
   {%- set identifier = model['name'] -%}
-  {%- set non_destructive_mode = flags.NON_DESTRUCTIVE == True -%}
+  {%- set tmp_identifier = identifier + '__dbt_tmp' -%}
+  {%- set non_destructive_mode = (flags.NON_DESTRUCTIVE == True) -%}
+  {%- set existing = adapter.query_for_existing(schema) -%}
+  {%- set existing_type = existing.get(identifier) -%}
 
+  -- setup
   {% if non_destructive_mode -%}
-    {%- if adapter.already_exists(schema, identifier) -%}
-        create temporary table {{ identifier }}__dbt_tmp {{ dist }} {{ sort }} as (
+    {% if existing_type == 'table' -%}
+      {{ adapter.truncate(identifier) }}
+    {% elif existing_type == 'view' -%}
+      {{ adapter.drop(identifier, existing_type) }}
+    {%- endif %}
+  {%- endif %}
+
+  {{ run_hooks(pre_hooks) }}
+
+  -- build model
+  {% statement capture_result -%}
+    {%- if non_destructive_mode -%}
+      {%- if adapter.already_exists(schema, identifier) -%}
+        create temporary table {{ tmp_identifier }} {{ dist }} {{ sort }} as (
           {{ sql }}
         );
 
         {% set dest_columns = adapter.get_columns_in_table(schema, identifier) %}
         {% set dest_cols_csv = dest_columns | map(attribute='quoted') | join(', ') %}
 
-        insert into {{ schema }}.{{ identifier }} ({{ dest_cols_csv }})
-        (
+        insert into {{ schema }}.{{ identifier }} ({{ dest_cols_csv }}) (
           select {{ dest_cols_csv }}
           from "{{ identifier }}__dbt_tmp"
         );
-    {%- else -%}
+      {%- else -%}
         {{ dbt__simple_create_table(schema, identifier, dist, sort, sql) }}
+      {%- endif -%}
+    {%- else -%}
+      {{ dbt__simple_create_table(schema, tmp_identifier, dist, sort, sql) }}
     {%- endif -%}
-  {%- elif non_destructive_mode -%}
-    {{ dbt__simple_create_table(schema, identifier, dist, sort, sql) }}
+  {%- endstatement %}
+
+  {{ run_hooks(post_hooks) }}
+
+  -- cleanup
+  {% if non_destructive_mode -%}
+    -- noop
   {%- else -%}
-    {% set tmp_identifier = identifier + '__dbt_tmp' %}
-    {{ dbt__simple_create_table(schema, tmp_identifier, dist, sort, sql) }}
+    {%- if existing_type is not none -%}
+      {{ adapter.drop(identifier, existing_type) }}
+    {%- endif %}
+
+    {{ adapter.rename(tmp_identifier, identifier) }}
   {%- endif %}
+
+  {{ adapter.commit() }}
```

Comment:
this is so so so good. This logic is necessarily kind of convoluted but I think it belongs exactly here in this file and not off in Python-land somewhere